feat: add exponential retries on merge errors

ricardoatsouza opened the pull request:
This Pull Request fulfills the following requirements:

- [x] The commit message follows our guidelines.
- [x] Tests for the changes have been added if needed.
- [x] Does not affect documentation or it has been added or updated.

<!--
Example: Resolves #1234

This allows PRs to be resolved automatically once the PR is merged to a base
branch. The following line should be removed if the PR does not affect any open
issues.
-->

# Resolves

Add an exponential retry on merge error, basically retrying up to three times (1 second after the first failure, then 4 seconds later and, finally, again 9 seconds later). If all three fail, an error is thrown.

viestat commented:
Why is the `dist`folder removed? As far as I know, GitHub Actions do not work with TypeScript. Did something changed?

ricardoatsouza commented:
> Looks good the only thing I am missing is the README changes, there we can describe how to configure the number of retires.

`README.MD` updated in c97bfc0.

JrSchild commented on src/common/merge.ts:
```diff
@@ -7,30 +7,53 @@ import {
+  details: {
+    numberOfRetries: number;
+    trial: number;
+  } & PullRequestDetails,
```
I would like to see this intersection definition swapped, so first the `PullRequestDetails`, then the extension.

ricardoatsouza replied:
Swapped in d1fdb79

JrSchild commented on src/eventHandlers/checkSuite/index.spec.ts:
```diff
@@ -334,8 +336,68 @@ describe('check Suite event handler', (): void => {
+    nock('https://api.github.com')
+      .post('/graphql')
+      .reply(OK, {
+        data: {
```
Can you put the contents of this in a generator?

ricardoatsouza replied:
I can, but I feel your request should go into a different PR. Every test has a mocked endpoint and their `data` content can be moved into a generator. I would rather prefer to have a dedicated PR to it.

JrSchild replied:
Ok sure, it's just to dry things out as I saw it was quite verbose and duplicated.

JrSchild commented on src/eventHandlers/checkSuite/index.spec.ts:
```diff
@@ -334,8 +336,68 @@ describe('check Suite event handler', (): void => {
+    const mergeSpy = jest
+      .spyOn(merge, 'merge')
+      .mockImplementation()
+      .mockRejectedValue(new Error('Error when merging'));
```
We shouldn't be mocking internal implementations somewhere in the middle of the call chain of the test. Instead we should always strive to mock the outgoing call.

ricardoatsouza replied:
Why not? Given the current use case, it doesn't matter which one, the outgoing call or a function, is mocked. The end result is the `merge` function failing with an error.

JrSchild replied:
The problem is that currently the mock is independent of the implementation. It is overriding the behavior and interface of the function. If the interface or implementation of the merge function changes, the tests wouldn't fail because it will consistently provide the same results because it is mocked.

What we want is that if the implementation or interface of the merge function changes, the test fail, so that we can correctly update the mock. So by mocking the response of the request (e.g. with nock), the merge function reacts to this change. Any change to the implementation of merge function will be picked up by the tests here in `checkSuite/index.spec.ts`.

ricardoatsouza replied:
I don't fully agree. The test is focused on the retries and backoff strategy. These can only happen when the `merge` function fails, independent of its implementation or interface. From the perspective of this test, the `merge` function is a "black box": it simply needs to fail. 

Any change in the implementation or interface of the `merge` function should not affect this particular test because its implementation is not what is being tested. Of course, that being said, there are other tests that cover and validates the implementation of said function.

JrSchild replied:
The reason it is tested entirely top down for when that function passes is the same reason why it should be tested top down for when it fails. The contract of the merge function can change at any time in which case the test should be failing here. Because the merge function could be throwing, does not mean that it will in the future as the implementation might change. A hidden implementation assumes that it can throw, which is just an assumption. The only way to be sure is to test it top down.

ricardoatsouza replied:
The `merge` function is being called only by `mergeWithRetry`, so a contract change wouldn't fail the test. An implementation change, such as not throwing an error, in the `merge` function requires a change in `mergeWithRetry`, invalidating the test as well. However, that's not explicit. Ok, I see what you mean. Thanks. Fixed in e28bc57.

JrSchild commented on src/eventHandlers/checkSuite/index.spec.ts:
```diff
@@ -334,8 +336,68 @@ describe('check Suite event handler', (): void => {
     expect(warningSpy).toHaveBeenCalled();
   });
+
+  it('retries up to two times before failing', async (): Promise<void> => {
```
A test is missing where you assert the backoff strategy.

ricardoatsouza replied:
I have added one more check in the tests (see e849d10) to check whether the second retry will happen in the logged time. But I am not sure if this is what you mean. Otherwise, care to explain?

JrSchild replied:
Yes that's fine actually

ricardoatsouza commented on README.md:
```diff
@@ -142,6 +142,24 @@ steps:
+
+In case the merge action fails, by default it will automatically be retried up
+to three times using an exponential backoff strategy. This means, the first
+retry will happen 1 second after the first failure, while the second will happen
```
A lot can happen in 1 second. Merge me action is simply waiting for Github to release `master` to merge and that can happen in this period. So, in 1 second it might be possible to merge already, speeding up the whole process.

ricardoatsouza commented on src/common/merge.ts:
```diff
@@ -7,30 +7,53 @@ import {
+  octokit: ReturnType<typeof getOctokit>,
+  details: {
+    numberOfRetries: number;
+    trial: number;
```
They mean the same in this context, but sure. Done in 29742be

ricardoatsouza commented on src/common/merge.ts:
```diff
@@ -7,30 +7,53 @@ import {
+export const mergeWithRetry = async (
+  octokit: ReturnType<typeof getOctokit>,
+  details: {
+    numberOfRetries: number;
```
Done in 7f874f6

JrSchild commented on src/eventHandlers/checkSuite/index.spec.ts:
```diff
@@ -334,8 +336,69 @@ describe('check Suite event handler', (): void => {
+    } catch (error) {
+      expect(error).toBeInstanceOf(Error);
+      expect(error.message).toStrictEqual('Error when merging');
+      expect(mergeWithRetrySpy).toHaveBeenCalledTimes(3);
```
This assertion doesn't seem necessary to me as it is part of the internal implementation.

JrSchild replied:
@merlinnot Getting stuck here with @ricardoatsouza. WDYT?

merlinnot replied:
Hm, that seems right - we should be able to replace entire implementation without any tests breaking, as long as the behavior presented externally remains the same.

ricardoatsouza replied:
@JrSchild pointed out that nock is already ensuring the endpoint to be called `3` times via `times(3)`. That's true, however, for example, if that number is changed to `4`, nock will fail with `Pending mocks detected`, which does not directly reflects the input given for number of retries.

My arguments are:
 - The number of `times` in nock is not related to the number of retries. If the implementation of the `merge` function changes, for example, by calling the endpoint two times instead of one, the number in nock will not anymore coincide with the number of retries.
 - This assertion, though an internal implementation, reflects what is expected of the functionality of the feature: to retry a certain number of times.

merlinnot replied:
The test will fail if it doesn't behave correctly, which is what we want.

> The number of times in nock is not related to the number of retries. If the implementation of the merge function changes, for example, by calling the endpoint two times instead of one, the number in nock will not anymore coincide with the number of retries.

The same can be said for the function: it can call itself two times at the same moment if you change the implementation. I don't really see how it changes anything.

What's more, we'll most likely refactor it as soon as our backoff implementation is extracted into a package and open sourced. We currently have an implementation for it used in multiple internal libraries and applications, which is unfortunately copy-pasted.

ricardoatsouza replied:
> The test will fail if it doesn't behave correctly, which is what we want.

The test will fail in both cases. My reasoning is towards what is being tested and the reason for a test to fail. What is being tested is the number of calls of a certain function which has to match with the number of retries given as input. The number of cals towards an endpoint is conveniently the same number, but does not mean the same thing.

> What's more, we'll most likely refactor it as soon as our backoff implementation is extracted into a package and open sourced. We currently have an implementation for it used in multiple internal libraries and applications, which is unfortunately copy-pasted.

What does this means in terms of this PR? To proceed and merge once the comments are addressed (and refactor in the future), or to wait for the backoff implementation to be open sourced and then implement it here?

merlinnot replied:
> What does this means in terms of this PR? To proceed and merge once the comments are addressed (and refactor in the future), or to wait for the backoff implementation to be open sourced and then implement it here?

That is an argument towards testing external behavior exclusively, as we'll be replacing the implementation. That's always the case, good tests are not dependent on specific implementation, they only test the behavior. Here, we're clearly testing this concrete implementation, which does not allow us to refactor it without refactoring tests. That should never be the case.

ricardoatsouza replied:
This line checks whether the already exported function is actually being called certain number of times, which is to me behavior related, and not a specific part of the implementation. Depending on what is being refactored, the output will still be the same.

JrSchild replied:
The behavior is already being tested with nock. We have made sure that the endpoint is being called with, and fails exactly `n` times. That allows us to change the implementation any way we want without changing the tests. So as @merlinnot pointed out, there should be no need to test the internal implementation because we have already covered the behavior.

ricardoatsouza replied:
> So as @merlinnot pointed out, there should be no need to test the internal implementation because we have already covered the behavior.

I don't agree. The fact that the endpoint is being called exactly `n` times is more of a _coincidence_. The implementation could change in a way that makes the endpoint be called `n + x` times, which would not only make the test fail, but do not reflect the number `n` of retries.

ricardoatsouza replied:
Removed in d0b59dc

ricardoatsouza commented:
@JrSchild @merlinnot 

Added parameter to speed up tests: ef0cedc

Retrying only on the [reported error](https://github.com/ridedott/merge-me-action/issues/458): 7cf6771 ( appreciate feedback on this one)

ricardoatsouza commented:
> @ricardoatsouza can you please revert #ef0cedc3674f849c0f0120c0d0d990d2f28537f1. You are exposing an internal implementation by changing the interface, but the option is never used in production. Also starting the backoff from 100ms instead is definitelly not the correct way to "speed up the test". Any timers that run longer than the an immediate interval should be avoided as much as possible. See our (internal) firestore-extensions package (retries.spec.ts) on how to properly test it.

It's not the wrong way either. Regardless:
 - Removed in 2032586
 - Using mock in aabf783
 
> Can you add a test that makes sure backoff does not happen if shouldRetry does not match the error?

Added in d5400ab
